Add ocean ability to output max and min quantities in timeSeriesMonthly

mark-petersen opened the pull request:
This adds the ability to output maximum and minimum observed quantities for each month. It utilizes the current timeseries AM functionality.

[BFB]
[NML]

mark-petersen commented:
This PR is identical to #2884 but for maint-1.0. It currently just updates the MPAS-Dev maint-1.0 branch to include the MPAS commits in https://github.com/MPAS-Dev/MPAS-Model/pull/209.

@jonbob I tried to cherry-pick the commits here:
https://github.com/E3SM-Project/E3SM/pull/2884/commits
but got conflicts, so I thought it would be best to let you do that. Thanks!

vanroekel commented on components/mpas-ocean/bld/build-namelist:
```diff
@@ -936,10 +936,12 @@ if ($ocn_bgc eq 'eco_only' || $ocn_bgc eq 'eco_and_dms' || $ocn_bgc eq 'eco_and_
 } else {
         add_default($nl, 'config_ecosys_atm_co2_option', 'val'=>"diagnostic");
         add_default($nl, 'config_ecosys_atm_alt_co2_option', 'val'=>"diagnostic");
+        add_default($nl, 'config_ecosys_atm_alt_co2_use_eco', 'val'=>".true.");
```
is it expected to change ecosystem flags in this PR?

jonbob replied:
No, I didn't think so either. That change was in the mpas-ocean Registry_ecosys.xml file and was picked up by the scripts that auto-generate part of the corresponding E3SM bld scripts. We'll have to involve @mark-petersen in fixing this.

mark-petersen replied:
Strange. I did a white space clean-up on https://github.com/MPAS-Dev/MPAS-Model/commit/acf23bbd4ea6de99c250702dccb4095b58900e79 (search for  `config_ecosys_atm_alt_co2_use_eco`)  where I changed spaces to tabs in front of that flag definition, which is what it is supposed to be. The scripts apparently only process the flags with the correct tab formatting, which is why it is shown as an addition now.

I think that the MPAS-Ocean file 
`src/core_ocean/tracer_groups/Registry_ecosys.xml`
is now correct, and the above changes should have been in the build-namelist file all along. Since it was not, it was always using the default value of true before for E3SM runs.

@maltrud do you agree with the above false/true settings?

We could either leave these two lines in this PR (and the one below in `build-namelist-section`), or make a separate E3SM PR. I'm fine with either.

mark-petersen replied:
Whoops, I missed that this is the one for maint-1.0. It probably doesn't matter either way, then, since that branch is not for BGC sims.

jonbob replied:
@vanroekel - can you please look at the changes and add your review? I'd like to get this committed before July 8 when the high-res run may restart....

vanroekel commented on components/mpas-ocean/bld/build-namelist-section:
```diff
@@ -483,6 +483,7 @@ add_default($nl, 'config_use_debugTracers_ttd_forcing');
 add_default($nl, 'config_use_ecosysTracers');
 add_default($nl, 'config_ecosys_atm_co2_option');
 add_default($nl, 'config_ecosys_atm_alt_co2_option');
+add_default($nl, 'config_ecosys_atm_alt_co2_use_eco');
```
same thing, is this expected?

vanroekel commented on components/mpas-ocean/bld/namelist_files/namelist_defaults_mpaso.xml:
```diff
@@ -233,6 +206,33 @@
+<config_min_diff_tanh>1.0e-5</config_min_diff_tanh>
+<config_zMid_tanh>-100</config_zMid_tanh>
+<config_zWidth_tanh>100</config_zWidth_tanh>
+
```
should this be changed?

jonbob replied:
none of these settings are changed -- the same things are removed above. But the Registry file must have been shuffled and the auto-generation scripts just follow that for the ordering of the namelists

vanroekel commented on components/mpas-ocean/bld/namelist_files/namelist_defaults_mpaso.xml:
```diff
@@ -721,6 +722,34 @@
+<config_AM_timeSeriesStatsMonthlyMax_backward_output_offset>'00-01-00_00:00:00'</config_AM_timeSeriesStatsMonthlyMax_backward_output_offset>
+
+<!-- AM_timeSeriesStatsMonthlyMin -->
+<config_AM_timeSeriesStatsMonthlyMin_enable>.false.</config_AM_timeSeriesStatsMonthlyMin_enable>
```
can we make the default true?

vanroekel commented on components/mpas-ocean/bld/namelist_files/namelist_defaults_mpaso.xml:
```diff
@@ -721,6 +722,34 @@
 <config_AM_timeSeriesStatsClimatology_backward_output_offset>'00-03-00_00:00:00'</config_AM_timeSeriesStatsClimatology_backward_output_offset>
 
+<!-- AM_timeSeriesStatsMonthlyMax -->
+<config_AM_timeSeriesStatsMonthlyMax_enable>.false.</config_AM_timeSeriesStatsMonthlyMax_enable>
```
can we make the default true?

jonbob replied:
yes, of course. I'll make that change

jonbob commented:
@mark-petersen - what's the status of this? @vanroekel needs this tracked down before we can restart the high-res run

mark-petersen commented:
I rebased on maint1.0, and removed those extra `config_ecosys_atm_alt_co2_use_eco` flags from the namelist. If you click on the `Files changed` tab and go to the bottom and click on the MPAS submodule files, you can see the only MPAS changes are those associated with this PR.

@jonbob I think this answers your question, and it is ready to merge.

jonbob commented:
@mark-petersen - I think we would like to understand where those flags came from, not simply remove them. We just need to be sure we have the correct codebase for continuing the high-res runs

mark-petersen commented:
The `config_ecosys_atm_alt_co2_use_eco` was added to the MPAS code with commit https://github.com/MPAS-Dev/MPAS-Model/commit/4aed07aa44c08852b5f379446d4b7de4117646f7 which was merged into MPAS ocean/develop on Apr 2 2018 with https://github.com/MPAS-Dev/MPAS/pull/1535. So it is supposed to be in maint-1.0. I don't know why it was missing in the E3SM namelists, but I don't think it is worth chasing any further.

vanroekel commented on components/mpas-ocean/cime_config/buildnml:
```diff
@@ -893,6 +983,38 @@ if ( -e "$CASEROOT/SourceMods/src.mpaso/$STREAM_NAME" ) {
+        if ( $OCN_GRID =~ m/^oRRS1/ ) {
+        	print $stream_file '        io_type="pnetcdf,cdf5"' . "\n";
+        } else {
+		print $stream_file '        io_type="' . "$OCN_PIO_TYPENAME" . '"' . "\n";
```
I think we can delete this check for grid.  These are 2-d cell centered fields, so they won't violate size requirements even for the 18to6

vanroekel commented on components/mpas-ocean/cime_config/buildnml:
```diff
@@ -838,7 +838,97 @@ if ( -e "$CASEROOT/SourceMods/src.mpaso/$STREAM_NAME" ) {
+        if ( $OCN_GRID =~ m/^oRRS1/ ) {
+                print $stream_file '        io_type="pnetcdf,cdf5"' . "\n";
+        } else {
+                print $stream_file '        io_type="' . "$OCN_PIO_TYPENAME" . '"' . "\n";
```
Same as below, I think we can remove the grid check
